python: add extra documentation to kvs.py module

chu11 opened the pull request:
Added a bit more in there.  Didn't add it for everything, sorta picked and choosed, notably not adding docs for the direct access to C-bindings stuffs b/c that stuff is mostly just used by `KVSDir`.  The readthedocs shouldn't be quite so bare now.

garlick commented on src/bindings/python/flux/kvs.py:
```diff
@@ -175,13 +189,16 @@ def __init__(self, flux_handle=None, path=".", handle=None):
         return p_str.decode("utf-8")
 
     def exists(self, name):
+        """Evaluate if key exists in the KVS"""
```
Should this say "in the KVS directory"?

chu11 replied:
gahhhhhh, I think I am right at the moment, BUT .... this is another "initial path" bug ... it should be "KVS directory"  .... I will push a separate fix.

```
diff --git a/src/bindings/python/flux/kvs.py b/src/bindings/python/flux/kvs.py
index ded222526..38a91814a 100644
--- a/src/bindings/python/flux/kvs.py
+++ b/src/bindings/python/flux/kvs.py
@@ -200,7 +200,7 @@ class KVSDir(WrapperPimpl, abc.MutableMapping):
 
     def exists(self, name):
         """Evaluate if key exists in the KVS"""
-        return exists(self.fhdl, name)
+        return exists(self.fhdl, self._path + name)
 
     def __getitem__(self, key):
         try:
diff --git a/t/python/t0005-kvs.py b/t/python/t0005-kvs.py
index 20efd9a62..344931aeb 100755
--- a/t/python/t0005-kvs.py
+++ b/t/python/t0005-kvs.py
@@ -90,8 +90,15 @@ class TestKVS(unittest.TestCase):
 
     def test_exists_true(self):
         flux.kvs.put(self.f, "thing", 15)
+        flux.kvs.put_mkdir(self.f, "existsubdir")
+        flux.kvs.put(self.f, "existsubdir.thing2", 14)
         flux.kvs.commit(self.f)
         self.assertTrue(flux.kvs.exists(self.f, "thing"))
+        self.assertTrue(flux.kvs.exists(self.f, "existsubdir.thing2"))
+
+        with flux.kvs.get_dir(self.f, "existsubdir") as kd:
+            self.assertTrue(flux.kvs.exists(self.f, "thing2"))
```

garlick reviewed (commented):
Despite the C heavy nature of the KVS API in flux-core, I think we're likely to have more python users over time, so focusing documentation efforts on the python API might be a good thing?

So perhaps we could get a bit more detail here?

chu11 commented:
> Despite the C heavy nature of the KVS API in flux-core, I think we're likely to have more python users over time, so focusing documentation efforts on the python API might be a good thing?

That's a fair point.  I choose to skip also because of https://github.com/flux-framework/flux-core/issues/5327, i.e. I'm not entirely sure what this should eventually look like.

But can definitely add more, it'll make documentation easier in the future down the road too.

chu11 commented:
re-pushed

- added way more documentation per comments above
- along the way realized some really old comments in `libkvs/kvs.h` were out of date, fixed that up
- along the way realized there were no coverage for commit flags, added some extra coverage
- realized there's an issue (#5333) added a temporary workaround for it in the tests

maybe could split out those testsuite commits into a new PR, but they are pretty tiny

grondo reviewed (approved):
LGTM!
